### ISSUE TITLE
Post-process data after normalization

### ISSUE THREAD

**Issue by dferber90** [paularmstrong/normalizr#206]
# Problem
In normalizr v2 we were able to use the `assignEntity` function to modify data after it has been normalized. In v3 this functionality went away and we can only pre-process it. Would you accept a PR that adds functionality which lets us modify the entities after they have been normalized (but before they are added into the entities of the result)?

## Use case
Our use-case is very specific. I'll try to dumb it down: We need to remove some parts of each entity before the normalization and add it back in after the normalization. We can't keep it in during the normalization because then it is not possible to create a proper schema.

<details>
<summary>The nasty details, feel free to skip them because they can cause headaches (left this here mostly for my team members)</summary>

A product we get from the API (two possibilities here):
```js
// with an expanded category reference
{
  id: 'product-1',
  name: 'Sponge',
  category: { typeId: 'category', id: 'category-1', obj: { id: 'category-1', /* other category data */ } }
}

// with a non-expanded category reference
{
  id: 'product-1',
  name: 'Sponge',
  category: { typeId: 'category', id: 'category-1' }
}
```

In the first case we want the resulting entities to be:

```js
{
  results: 'product-1',
  entities: {
    products: {
      'product-1': { id: 'product-1', name: 'Sponge', category: { typeId: 'category', id: 'category-1' } },
    },
    categories: {
      'category-1': { id: 'category-1', /* other category data */  }
    }
  }
}
```

In the second case it would be 

```js
{
  results: 'product-1',
  entities: {
    products: {
      'product-1': { id: 'product-1', name: 'Sponge', category: { typeId: 'category', id: 'category-1' } },
    },
    categories: {
      'category-1': { id: 'category-1' }
    }
  }
}
```

Notice how we'd need to define the category schema on the `obj`? But what would happen if there is no `obj` like in the second case? Another problem here is that the category is contained within `obj`, but we can't define the schema there because then the reference after the normalization would end up in the wrong spot (`{ typeId: 'category', id: 'category-1', obj: 'category-1' }`, but it should be `{ typeId: 'category', id: 'category-1' }`. 

The solution we'd like to use is to preprocess the data to be able to transform it:

```js
// before preprocessing
{
  id: 'product-1',
  name: 'Sponge',
  category: { typeId: 'category', id: 'category-1' }
}

// after preprocessing (what the normalization will see)
{
  id: 'product-1',
  name: 'Sponge',
  category: { id: 'category-1' }
}
```

Now we can just define category to be a reference.

And then we can post-process the data and re-add the intermediate reference description:

```js
{
  id: 'product-1',
  name: 'Sponge',
  category: { typeId: 'category', id: 'category-1' }
}
```

This solution requires a preprocessor and a postprocessor. We'd need to generate these functions by passing in descriptions of which fields to add/remove expansions from (and what their `typeId` is called).

Thank you for taking the time to read/skim through all this.
</details>

## Possible solution

We had this working earlier by transforming the data before passing it to normalizr and then modifying the output using `assignEntity`. But now that `normalizr` uses `processEntity` we could get rid of our preprocessing system, but then we'd have to write our own post-processing system.

If I understand this correctly, what happened is that the release of v3 moved the processor from after the normalization to before. I think having both possibilities would give the most power and satisfy our use case. What was the motivation behind changing it?

I propose adding a `postprocessStrategy` to the options. Would that make sense? Would it cause a lot of work? We'd be up for trying to open a PR adding this functionality in case you approve.

**Comment by paularmstrong** [comment]
I'm resistant to this yet. (See #199 for another reference and recommended approach. tl;dr: your reducer should be handling things)

It seems like you're normalizing things a little oddly. I would expect things to be more flat.  You're trying to normalize `category` an extra level deep *sometimes*. Instead, why not normalize it at the top level `category` key in `product`? Once this is done, your reducer can figure out what to do with complete/incomplete entities.

```js
const data = {
  expanded: {
    id: 'product-1',
    name: 'Sponge',
    category: { typeId: 'category', id: 'category-1', obj: { id: 'category-1', tacos: 'yum' } }
  },
  bare: {
    id: 'product-1',
    name: 'Sponge',
    category: { typeId: 'category', id: 'category-1' }
  }
};

const category = new schema.Entity('category', {}, {
  processStrategy: (value, parent, key) => {
    const { obj, ...nonObjValue } = value
    return {
      ...nonObjValue,
      ...obj
    };
  }
});
const product = new schema.Entity('products', { category });
```

Using your "expanded" example data:

```js
normalize(data.expanded, product);
// -> 
{
  "entities": {
    "category": {
      "category-1": {
        "typeId": "category",
        "id": "category-1",
        "tacos": "yum"
      }
    },
    "products": {
      "product-1": {
        "id": "product-1",
        "name": "Sponge",
        "category": "category-1"
      }
    }
  },
  "result": "product-1"
}
```

And using your "bare" example data:

```js
normalize(data.bare, product);
// ->
{
  "entities": {
    "category": {
      "category-1": {
        "typeId": "category",
        "id": "category-1"
      }
    },
    "products": {
      "product-1": {
        "id": "product-1",
        "name": "Sponge",
        "category": "category-1"
      }
    }
  },
  "result": "product-1"
}
```

**Comment by dferber90** [comment]
*Thank you for the fast and detailed response. While writing everything down I realized that while postprocessing would be helpful we can probably get around not having it without any real downsides by using a middleware.*


Unfortunately your answer slightly missed the point since the reference descriptions are not part of the entities (so no `typeId` in `category`). To clear it up: Whenever one entity contains another entity (like products contains a category in `category`) the other entity is not included directly, but an intermediate reference object is there (`{ typeId, id, obj }`) and `obj` contains the actual entity (the category in this case).

<details>
<summary>Going through what we want to do step-by-step:</summary>
What we get from the API
```js
// with an expanded category reference
{
  id: 'product-1',
  name: 'Sponge',
  category: { typeId: 'category', id: 'category-1', obj: { id: 'category-1', taco: 'yum' } }
}

// with a non-expanded category reference
{
  id: 'product-1',
  name: 'Sponge',
  category: { typeId: 'category', id: 'category-1' }
}
```

After the preprocessing (removing reference descriptions)
```js
// with an expanded category reference
{
  id: 'product-1',
  name: 'Sponge',
  category: { id: 'category-1', taco: 'yum' } } // no typeId, just the stuff that was inside "obj"
}

// with a non-expanded category reference
{
  id: 'product-1',
  name: 'Sponge',
  category: { id: 'category-1' } // we don't know anything about the category, just the id
}
```

Now `normalizr` can do its thing without having to worry about our structure because we removed the oddities in the preprocessing.

```js
// with an expanded category reference
{
  entities: {
    products: { 'product-1': { id: 'product-1', name: 'Sponge', category: 'category-1' } },
    categories: { 'category-1': { id: 'category-1', taco: 'yum' } },
  }
}

// with a non-expanded category reference
{
  entities: {
    products: { 'product-1': { id: 'product-1', name: 'Sponge', category: 'category-1' } },
    categories: { 'category-1': { id: 'category-1' } },
  }
}
```

And now to the point where the postprocessing should add the oddities back:

```js
// with an expanded category reference
{
  entities: {
    products: { 'product-1': { id: 'product-1', name: 'Sponge', category: { typeId: 'category', id: 'category-1' /* no "obj" here since that is normalized now */ } } },
    categories: { 'category-1': { id: 'category-1', taco: 'yum' } },
  }
}

// with a non-expanded category reference
{
  entities: {
    products: { 'product-1': { id: 'product-1', name: 'Sponge', category: { typeId: 'category', id: 'category-1' } } },
    categories: { 'category-1': { id: 'category-1' } },
  }
}
```

</details>

To sum it up, we are roughly doing this:
`addReferenceDescriptions(normalize(removeReferenceDescriptions(data), schema))`

What this issue and the other one boil down to is that whenever we want to transform the data before/after it has been normalized, we need to create a system that knows about the entity relationships. But we already have that information in `normalizr`'s schema, so we'd actually have to describe the data and the relationships once for preprocessing, once for normalizr and once for postprocessing. And to transform them, we need to walk through the data two times (pre, normalizr) and then transform the entities of the `normalizr` output.

If normalizr had post-processing we:
- would be able to walk through the data only once
- could have the whole entity relationship descriptions in a single place

Thank you for taking the time to deal with this. Absolutely no pressure on adding it as we're not sure about this ourselves.

**Comment by kolpav** [comment]
You can solve almost anything with reducer. By having post and pre normalization hooks you could tightly couple some processing logic with your entities and it just feels right for some use cases e.g. omitting properties with pre hook and converting to immutablejs record with post hook. I think every argument against post hook is equally valid for pre hook yet we have it. I think we should have both or none of them.

But I understand you can't please everyone and add every api people wish for. Maybe post hook is anti-pattern and I just don't see it.

**Comment by tugorez** [comment]
@paularmstrong first of all: thank you very much for such a nice library :) , second I don't see why this is "bad" and it should be handled by a reducer. 
I'm trying to do the same as @kolpav  in  https://github.com/paularmstrong/normalizr/issues/199 and it just feels the right place (and not in the reducer) to cast the entities to Immutable.Records.

**Comment by tugorez** [comment]
@paularmstrong @kolpav @dferber90, I did the change https://github.com/tugorez/normalizr 

you could see the usage here https://github.com/tugorez/normalizr/blob/8a040f02e9e4963cde2706f8a724347d705d2927/src/schemas/__tests__/Entity.test.js#L119

what do you think ?
